tests/test_quoting.py

```diff
@@ -455,3 +457,82 @@ def test_quoter_path_with_plus(quoter):
 def test_unquoter_path_with_plus(unquoter):
     s = "/test/x+y%2Bz/:+%2B/"
     assert "/test/x+y+z/:++/" == unquoter(unsafe="+")(s)
+
+
+@example(
+    # quoter=_PyQuoter,
+    # unquoter=_PyUnquoter,
+    text_input="0",
+    safe="",
+    unsafe="0",
+    protected="",
+    qs=False,
+    requote=False,
+)
+@settings(suppress_health_check=(HealthCheck.function_scoped_fixture,))
+@given(
+    text_input=st.text(),
+    safe=st.text(alphabet=st.characters(max_codepoint=127)),
+    unsafe=st.text(),
+    protected=st.text(alphabet=st.characters(max_codepoint=127)),
+    qs=st.booleans(),
+    requote=st.booleans(),
+)
+def test_quote_unquote_parameter(
+    quoter: _PyQuoter,
+    unquoter: _PyUnquoter,
+    safe: str,
+    unsafe: str,
+    protected: str,
+    qs: bool,
+    requote: bool,
+    text_input: str,
+) -> None:
+    quote = quoter(safe=safe, protected=protected, qs=qs, requote=requote)
+    unquote = unquoter(unsafe=unsafe, qs=qs)
+
+    text_quoted = quote(text_input)
+    assume(qs and all(unsafe_char not in text_quoted for unsafe_char in unsafe))
+    note(f"text_quoted={text_quoted !r}")
+    text_output = unquote(text_quoted)
+    assume(qs or all(unsafe_char not in text_output for unsafe_char in unsafe))
+    # assume(
+    #     qs and
+    #     (unsafe in text_quoted and text_input != text_output) or
+    #     (unsafe not in text_quoted)
+    # )
+
+    assert text_input == text_output
```

Comment:
There are cases where we have defaults for safe so we can't assume round trip for these

https://github.com/aio-libs/yarl/blob/d2892dadc1ba97626f3d89656374eef5b5598d7e/yarl/_quoting_py.py#L49
https://github.com/aio-libs/yarl/blob/d2892dadc1ba97626f3d89656374eef5b5598d7e/yarl/_quoting_py.py#L13

```
FAILED tests/test_quoting.py::test_quote_unquote_parameter_with_qs[c_quoter-py_unquoter] - AssertionError: assert '\x1f' == '%1F'
FAILED tests/test_quoting.py::test_quote_unquote_parameter_with_qs[py_quoter-py_unquoter] - AssertionError: assert '#' == '%23'
FAILED tests/test_quoting.py::test_quote_unquote_parameter_with_qs[py_quoter-c_unquoter] - AssertionError: assert '\x1f' == '%1F'
FAILED tests/test_quoting.py::test_quote_unquote_parameter_with_qs[c_quoter-c_unquoter] - AssertionError: assert '+' == '%2B'
```